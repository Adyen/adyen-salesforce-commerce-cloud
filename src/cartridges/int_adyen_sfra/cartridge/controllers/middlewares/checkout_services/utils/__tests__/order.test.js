jest.mock('../transaction');
jest.mock('../../helpers/index', () => ({
  hasAdyenPaymentMethod: jest.fn(),
}));
jest.mock('../payment');
jest.mock('../fraud');

const COHelpers = require('*/cartridge/scripts/checkout/checkoutHelpers');
const createOrder = require('../order');
const handlePaymentAuthorization = require('../payment');
const handleFraudDetection = require('../fraud');
const handleTransaction = require('../transaction');
const { hasAdyenPaymentMethod } = require('../../helpers/index');

let req;
let res;
let emit;
let next;
beforeEach(() => {
  jest.clearAllMocks();
  req = {
    locale: { id: 'nl_NL' },
    session: { privacyCache: { set: jest.fn() } },
    currentCustomer: {addressBook: true},
  };
  res = { json: jest.fn() };
  emit = jest.fn();
  next = jest.fn();
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('Order', () => {
  it('should go to next middleware if paymentMethod is not Adyen', () => {
    hasAdyenPaymentMethod.mockReturnValue(false);
    createOrder({}, { req, res, next }, emit);
    expect(next).toHaveBeenCalledTimes(1);
    expect(handleTransaction).toBeCalledTimes(0);
  });
  it('should return nothing on invalid transaction', () => {
    hasAdyenPaymentMethod.mockReturnValue(true);
    handleTransaction.mockReturnValue(false);
    const result = createOrder({}, { req, res, next }, emit);
    expect(result).toBeUndefined();
    expect(next).toHaveBeenCalledTimes(0);
  });
  it('should return json response with error when order is not created', () => {
    COHelpers.createOrder.mockReturnValue(false);
    hasAdyenPaymentMethod.mockReturnValue(true);
    handleTransaction.mockReturnValue(true);

    createOrder({}, { req, res, next }, emit);
    expect(res.json.mock.calls).toMatchSnapshot();
    expect(next).toHaveBeenCalledTimes(0);
    expect(emit).toBeCalledWith('route:Complete');
  });
  it('should return nothing when payment is not authorized', () => {
    handlePaymentAuthorization.mockReturnValue(false);
    COHelpers.createOrder.mockReturnValue(true);
    hasAdyenPaymentMethod.mockReturnValue(true);
    handleTransaction.mockReturnValue(true);

    const result = createOrder({}, { req, res, next }, emit);
    expect(result).toBeUndefined();
    expect(handleTransaction).toBeCalledTimes(1);
  });
  it('should return nothing when fraud detection is unsuccessful', () => {
    handleFraudDetection.mockReturnValue(false);
    handlePaymentAuthorization.mockReturnValue(true);
    COHelpers.createOrder.mockReturnValue(true);
    hasAdyenPaymentMethod.mockReturnValue(true);
    handleTransaction.mockReturnValue(true);

    const result = createOrder({}, { req, res, next }, emit);
    expect(result).toBeUndefined();
    expect(handlePaymentAuthorization).toBeCalledTimes(1);
  });
  it('should return nothing when there is an error while placing the order', () => {
    COHelpers.placeOrder.mockReturnValue({ error: true });
    handleFraudDetection.mockReturnValue(true);
    handlePaymentAuthorization.mockReturnValue(true);
    COHelpers.createOrder.mockReturnValue(true);
    hasAdyenPaymentMethod.mockReturnValue(true);
    handleTransaction.mockReturnValue(true);

    const result = createOrder({}, { req, res, next }, emit);
    expect(result).toBeUndefined();
    expect(handleFraudDetection).toBeCalledTimes(1);
  });
  it('should confirm order', () => {
    COHelpers.placeOrder.mockReturnValue({ error: false });
    handleFraudDetection.mockReturnValue(true);
    handlePaymentAuthorization.mockReturnValue(true);
    COHelpers.createOrder.mockReturnValue({ orderNo: 1, orderToken: 'token' });
    hasAdyenPaymentMethod.mockReturnValue(true);
    handleTransaction.mockReturnValue(true);

    createOrder({}, { req, res, next }, emit);
    expect(res.json.mock.calls).toMatchSnapshot();
  });
});
