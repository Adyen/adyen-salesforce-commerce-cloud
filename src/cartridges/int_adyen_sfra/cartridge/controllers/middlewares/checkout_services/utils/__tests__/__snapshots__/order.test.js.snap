// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order should confirm order 1`] = `
Array [
  Array [
    Object {
      "continueUrl": "[\\"Order-Confirm\\"]",
      "error": false,
      "orderID": 1,
      "orderToken": "token",
    },
  ],
]
`;

exports[`Order should return json response with error when order is not created 1`] = `
Array [
  Array [
    Object {
      "error": true,
      "errorMessage": "mocked_error.technical",
    },
  ],
]
`;
